servers/src/main/java/tachyon/worker/block/TieredBlockStore.java

```diff
@@ -0,0 +1,432 @@
+        mLockManager.unlockBlock(lockId);
+      }
+    } finally {
+      // If we fail to lock, the block is no longer in tiered store
+      mEvictionLock.readLock().unlock();
+    }
+  }
+
+  @Override
+  public void removeBlock(long userId, long blockId) throws IOException {
+    mEvictionLock.readLock().lock();
+    try {
+      long lockId = mLockManager.lockBlock(userId, blockId, BlockLockType.WRITE);
+      try {
+        removeBlockNoLock(userId, blockId);
+        for (BlockStoreEventListener listener : mBlockStoreEventListeners) {
+          listener.onRemoveBlockByClient(userId, blockId);
+        }
+      } finally {
+        mLockManager.unlockBlock(lockId);
+      }
+    } finally {
+      // If we fail to lock, the block is no longer in tiered store
+      mEvictionLock.readLock().unlock();
+    }
+  }
+
+  @Override
+  public void accessBlock(long userId, long blockId) {
+    for (BlockStoreEventListener listener : mBlockStoreEventListeners) {
+      listener.onAccessBlock(userId, blockId);
+    }
+  }
+
+  @Override
+  public void freeSpace(long userId, long availableBytes, BlockStoreLocation location)
+      throws IOException {
+    mEvictionLock.writeLock().lock();
+    try {
+      freeSpaceInternal(userId, availableBytes, location);
+    } finally {
+      mEvictionLock.writeLock().unlock();
+    }
+  }
+
+  @Override
+  public void cleanupUser(long userId) throws IOException {
+    mEvictionLock.readLock().lock();
+    List<TempBlockMeta> tempBlocksToRemove = mMetaManager.cleanupUser(userId);
+    mLockManager.cleanupUser(userId);
+    mEvictionLock.readLock().unlock();
+    Set<String> dirs = new HashSet<String>();
+    for (TempBlockMeta tempBlockMeta : tempBlocksToRemove) {
+      String fileName = tempBlockMeta.getPath();
+      try {
+        String dirName = CommonUtils.getParent(fileName);
+        dirs.add(dirName);
+      } catch (InvalidPathException e) {
+        LOG.error("Cannot parse parent dir of {}", fileName);
+      }
+      if (!new File(fileName).delete()) {
+        throw new IOException("Failed to cleanup userId " + userId + ": cannot delete file "
+            + fileName);
+      }
+    }
+    // Cleanup the user folder
+    for (String dirName : dirs) {
+      if (!new File(dirName).delete()) {
+        throw new IOException("Failed to cleanup userId " + userId + ": cannot delete directory "
+            + dirName);
+      }
+    }
+  }
+
+  @Override
+  public BlockStoreMeta getBlockStoreMeta() {
+    return mMetaManager.getBlockStoreMeta();
+  }
+
+  @Override
+  public void registerBlockStoreEventListener(BlockStoreEventListener listener) {
+    mBlockStoreEventListeners.add(listener);
+  }
+
+  private TempBlockMeta createBlockMetaNoLock(long userId, long blockId,
+      BlockStoreLocation location, long initialBlockSize) throws IOException {
+    if (mMetaManager.hasTempBlockMeta(blockId)) {
+      throw new IOException("Failed to create TempBlockMeta: blockId " + blockId + " exists");
```

Comment:
maybe `is already being written` rather than `exists`? The difference between this and the next error message is not too clear.